### PR TITLE
Add doc with hostpolicy APIs

### PR DESCRIPTION
**Pull request by elinor-fung**
- Add doc listing existing hostpolicy APIs
- Add proposed new hostpolicy export (for supporting new native hosting scenarios)

### CODE REVIEW

**Comment by elinor-fung** [comment]
For the existing APIs, I pretty much just looked for `SHARED_API` in `hostpolicy` and then looked at the implementations to add descriptions, so I'm sure there's room for improved descriptions.

The one I'm suggesting adding is at the bottom.

File: Documentation/design-docs/hosting-layer-apis.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+``` C
+int corehost_main_with_output_buffer(
+    const int argc,
+    const char_t *argv[],
+    char_t buffer[],
+    int32_t buffer_size,
+    int32_t *required_buffer_size)
+```
+
+Run a host command and return the output. `corehost_load` should have been called with the `host_command` set on the `host_interface_t`. This function operates in the hosting layer and does not actually run CoreCLR.
```
Comment (sdmaclea) [inline Documentation/design-docs/hosting-layer-apis.md]:
```suggestion
Run a host command and return the output. `corehost_load(init)` should have been called with `init->host_command` set. This function operates in the hosting layer and does not actually run CoreCLR.
```

File: Documentation/design-docs/hosting-layer-apis.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+* `argc` / `argv` - command-line arguments
+* `buffer` - buffer to populate with the output (including a null terminator)
+* `buffer_size` - size of `buffer` in `char_t` units
+* `required_buffer_size` - if `buffer` is too small, this will be populated with the minimum required buffer size
```
Comment (sdmaclea) [inline Documentation/design-docs/hosting-layer-apis.md]: If the buffer is `too small`, will buffer be:

- Empty
- Filled with head/beginning of output
- Filled with tail/end of output

if buffer is not too small, will this be unchanged?

File: Documentation/design-docs/hosting-layer-apis.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        context_handle instance,
+        size_t *count,
+        const char_t **keys,
+        const char_t **values);
+};
+```
+
+Contract for performing operations on an initialized host context.
+* `version` - version of the struct.
+* `instance` - opaque handle to the initialized host context.
```
Comment (sdmaclea) [inline Documentation/design-docs/hosting-layer-apis.md]: `initialized` once initialized isn't it too late to set properties?

Comment (sdmaclea) [reply Documentation/design-docs/hosting-layer-apis.md]:
```suggestion
* `instance` - opaque handle to the `corehost_context_contract` state.
```

Comment (vitek-karas) [reply Documentation/design-docs/hosting-layer-apis.md]: The design is to run everything up to the point of loading `coreclr`, without actually loading it. So We can still modify properties.

File: Documentation/design-docs/hosting-layer-apis.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+Run a host command and return the output. `corehost_load` should have been called with the `host_command` set on the `host_interface_t`. This function operates in the hosting layer and does not actually run CoreCLR.
+* `argc` / `argv` - command-line arguments
+* `buffer` - buffer to populate with the output (including a null terminator)
+* `buffer_size` - size of `buffer` in `char_t` units
+* `required_buffer_size` - if `buffer` is too small, this will be populated with the minimum required buffer size
+
+### .NET Core 3.0+
+
+``` C
+int corehost_get_coreclr_delegate(coreclr_delegate_type type, void **delegate)
```
Comment (vitek-karas) [inline Documentation/design-docs/hosting-layer-apis.md]: Part of the proposal should be that we move the delegate stuff over to the new context APIs. This is new in 3.0 so there's no reason to use the "old" system for it.

File: Documentation/design-docs/hosting-layer-apis.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+* `set_property` - function pointer for setting a property on the host context.
+  * `key` - key of the property to set.
+  * `value` - value of the property to set. If `nullptr`, the property is removed.
+* `get_properties` - function pointer for getting all properties on the host context.
+  * `count` - size of `keys` and `values`. If the size is too small, it will be populated with the required size. Otherwise, it will be populated with the size used.
+  * `keys` - buffer to populate with the property keys
+  * `values` - buffer to populate with the property values
+
+``` C
+int corehost_initialize_context(const host_interface_t *init, corehost_context_contract *context_contract)
```
Comment (vitek-karas) [inline Documentation/design-docs/hosting-layer-apis.md]: We also need some kind of `corehost_close_context` or similar to free the context (would be called from `hostfxr_shutdown`).

File: Documentation/design-docs/hosting-layer-apis.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        const char_t **value);
+    int (*set_property)(
+        context_handle instance,
+        const char_t *key,
+        const char_t *value);
+    int (*get_properties)(
+        context_handle instance,
+        size_t *count,
+        const char_t **keys,
+        const char_t **values);
```
Comment (vitek-karas) [inline Documentation/design-docs/hosting-layer-apis.md]: What about the "run" functions?
One to get a delegate and another one to run the app.

File: Documentation/design-docs/hosting-layer-apis.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+#### [Proposed] .NET Core 3.0+
+
+``` C
+typedef void* context_handle;
+
+struct corehost_context_contract
+{
+    size_t version;
+    context_handle instance;
+    int (*get_property)(
```
Comment (vitek-karas) [inline Documentation/design-docs/hosting-layer-apis.md]: Nit: I would call this `get_property_value` (and `set_property_value`).

File: Documentation/design-docs/hosting-layer-apis.md (hunk 6 above)
Comment (elinor-fung) [reply Documentation/design-docs/hosting-layer-apis.md]: doh.

File: Documentation/design-docs/hosting-layer-apis.md (hunk 2 above)
Comment (elinor-fung) [reply Documentation/design-docs/hosting-layer-apis.md]: Update to include that:
`buffer` is unchanged in the case that the size too small.
`required_buffer_size` is set to 0 in the case that the size is not too small.

File: Documentation/design-docs/hosting-layer-apis.md [hunk 8]
```diff
@@ -143,3 +164,10 @@ int corehost_initialize_context(const host_interface_t *init, corehost_context_c
 Initializes the host context. This calculates everything required to start CoreCLR (but does not actually do so).
 * `init` - struct defining how the host context should be initialized. If the host context is already initialized, this function will check if `init` is compatible with the active context.
 * `context_contract` - if initialization is successful, this is populated with the contract for operating on the initialized host context.
+
+``` C
+int corehost_close_context(corehost_context_contract *context_contract)
+```
+
+Closes the host context. Once closed, any operations targeting the closed context will return an error.
```
Comment (vitek-karas) [inline Documentation/design-docs/hosting-layer-apis.md]: I would not commit to returning sensible errors, unless we want to build a handle table and so on. I was only thinking of casting the handle as a pointer to a struct. So closed handles are effectively freed pointers.
But this is just a nit - we'll get there during implementation.

**Pull request merged**